### ISSUE TITLE
Multimedia player: Rework control labels to be more accessible

### ISSUE THREAD

**Issue by EricDunsworth** [wet-boew/wet-boew#9304]
This is a spinoff of #8947.

This issue's goal is to revamp the labels of WET's media player controls to be more accessible (and less confusing/annoying).

Here are some key comments from the previous issue to use as a starting point...

**@eburggraaf's reply (https://github.com/wet-boew/wet-boew/pull/8947#issuecomment-997026748):**
> Hello, I'm Erik, and I've been working with Ben and Shawn on testing the proposed controls for this player using a screen reader. From earlier posts in this thread, it looks like aria-labelledby is being used to override the actual button values. Please reconsider this. It's technically incorrect and It's creating a lot of problems from an accessibility standpoint.
> 
> Because of aria-labelledby, the name of the video appears to be the control’s name from a screen reader point of view. This isn’t allowed. The buttons and controls all need to be self-aware so they can be found in a screen search. For example, if I go to the top of the webpage and do a jaws find for the word play, jaws finds nothing, because there isn’t a control named play on the screen. When I do an insert f5 to get the controls list, all of the player controls are called “looking for a job”. In both of these cases, we need to see “play”, “mute”, “volume”, and “Closed captions”. This is also the case when I use item chooser on the mac.
> 
> As you can see, the handling of the tooltips that contain the actual names of the buttons in various screen reader and browser combinations varies wildly, with some tips on some configurations being detected while others not, or only when tabbing or only when arrowing. Ideally, we want the tooltip information displayed in the control name and that currently displayed in the control name in the tooltip. The tooltips appear to come from title attributes, an approach that I personally like, but I suppose only because it largely works in my favourite screen reader/browser combinations. It can't be used anyhow because of the below requirement for more than one player on a page.
> 
> If this is the way the page looks to dictation software, the player controls as they are will thwart that as well. I can see that the actual name of each button does properly reflect it's action. IE, the inner html of the buttons read, "Play", "Mute", etc. This will need to be tested if you decide to keep the aria-labelledby for some reason.
> 
> This lead to a further discussion among the testers because we have a requirement to support multiple instances of a player on a page. In order to maintain self-awareness, the buttons can't simply be called their actions as would be preferred. If that were the case, then we would have two or more "play", "mute"... And self-awareness goes out the window. Thus, we need to eligantly incorporate the name of the video into the button name.
> 
> We currently store the name of the video in the ID of the main player container. So clearly we have this variable that we can use. Here is a list of places we might be allowed to use that variable. In the innerHTML of the button, In the alt tag of an image in the innerHTML of the button, In the title attribute of the button that has no innerHTML.
> 
> In other words, in order to have multiple players on screen at once, the name of the video can not be missing from the buttons, but can in no way superceed the description of what action that button does. Examples of this would be: The title attribute of a button that has innerHTML, Aria-describedby, aria-labelledby.
> 
> Any of these will result in the button losing self-awareness.
> 
> We also discussed using regions, landmarks, and other grouping mechanisms to automatically speak the name of the video when a user enters a group. Unfortunately, that has the same inconsistancies as demonstrated by our spreadsheet of screen reader feedback, and still leaves the buttons unaware of how to describe themselves. So, no screen search, no clicking with dictation.
> 
> So to conclude on this, the variable containing the title of the video must be emplaced into either the innerHTML of the button alongside the button's action, or into an alt tag attached to an image in the innerHTML of the button alongside the button's action, or in a title attribute of a button that has no innerHTML alongside the button's action. Thus, the text of the button will visually and screenreaderically read "play video title" button, "mute video title" button... The buttons will be fully self-aware and detectable by access technologies accounting for more than one player on a page.
> 
> If we discard the requirement to have more than one player on a page, then we get a lot more flexibility in what we can do. That must be a conversation for another post though.

**@EricDunsworth's reply (https://github.com/wet-boew/wet-boew/pull/8947#issuecomment-1018897155):**
> @eburggraaf @shawnthompson What would you think about merging this PR "as-is" to get a quick win for #8939, then look into a bigger rework of the media player's control labels via a separate PR? As @duboisp mentioned, it doesn't look like this PR's change is degrading anything compared to what currently exists. All it does in practice is improve semantics.
> 
> FWIW I agree that the status quo with video names acting as prefixes for every media player control is super annoying and want to try sorting it out. But I wanted to start off by fixing the volume slider issue in what already exists before trying to rework things as a whole.
> 
> @eburggraaf I also had a few additional questions for you :).
> 
> **1.**
> 
> > The buttons and controls all need to be self-aware so they can be found in a screen search. For example, if I go to the top of the webpage and do a jaws find for the word play, jaws finds nothing, because there isn’t a control named play on the screen. When I do an insert f5 to get the controls list, all of the player controls are called “looking for a job”. In both of these cases, we need to see “play”, “mute”, “volume”, and “Closed captions”. This is also the case when I use item chooser on the mac.
> 
> Just wondering - do those screen reader features try appending `aria-describedby` values to the end of the control names they announce in the scenarios you mentioned (e.g. JAWS find/control list, VoiceOver item chooser, etc)?
> 
> For example, with the volume slider control (which uses a void `input` element):
> 
>     * Most screen readers currently announce the volume slider control in this order:
>       
>       * "[`aria-labelledby`] [slider description] [silder value] [`aria-describedby`] [JAWS: min/max values]"
> 
>     * Which comes out like this in some of your JAWS samples:
>       
>       * "[Looking for a Job] [left right slider] [95] [Volume] [min 0 max 100]"
> 
> 
> I believe the new kind of labelling structure you're proposing would cause the volume control to be announced like this:
> 
>     * "[Volume Looking for a Job] [left right slider] [95] [min 0 max 100]"
> 
> 
> Since users would probably prefer to hear the volume level before the video's name, I was thinking that maybe `aria-describedby` could come in handy to relegate the video's name to the end of the announcement. But only if that doesn't end up causing other issues.
> 
> **2.**
> 
> > In order to maintain self-awareness, the buttons can't simply be called their actions as would be preferred. If that were the case, then we would have two or more "play", "mute"... And self-awareness goes out the window. Thus, we need to eligantly incorporate the name of the video into the button name.
> 
> How would this apply to pages that only contain one instance of the multimedia player? Would you recommend trying to make the plugin capable of excluding the video name from control labels if it determines that there's only one video in the current page?

CC @eburggraaf @shawnthompson @duboisp

**Comment by EricDunsworth** [comment]
FWIW I started looking into addressing this issue a little while ago, but haven't reached the point to sending a PR for it as of yet. Hoping to be able to pick up on it later this month or in October.

@GormFrank @BrahimMahadi Would that interfere with your team's ongoing WCAG 2.2 assessments of WET's plugins? For example, #9399's goal is to assess the multimedia player. What would happen if I sent a PR that reworked its botched labelling in parallel to that review?

**Comment by EricDunsworth** [comment]
@CraigClark Did you mean that a11y review flagged the progress bar for *not* having a label?

As for where https://github.com/wet-boew/wet-boew/issues/9304#issuecomment-1255198206 stands... Never got a chance to follow-up on it :(, but I did have a stash of WIP changes lingering around for it somewhere. I'll see if I can dig it up and apply finishing touches to it in the near-term.

IIRC it was mostly-done and just needed a few finishing touches before sending out a PR. I don't believe it tried tackling progress bar labels though. I'll try to take it into consideration once I pick up on it.

**Comment by CraigClark** [comment]
@EricDunsworth Yes, flagged for not having labels
